### ISSUE TITLE
Use of current()

### ISSUE THREAD

**Issue by ttho** [ShapeChange/ShapeChange#13]
The assertion XPath expressions generated by ShapeChange make use of current(). However, this function is only defined in XSLT and not part of XPath.

It should simply use "." instead.

**Comment by jechterhoff** [comment]
The Schematron query bindings supported by ShapeChange in the conversion to Schematron are xslt and xslt2. That means that an XSLT based implementation is assumed (like the reference implementations).

Given this background (xslt based evaluation of the Schematron rules):

'.' and 'current()' are not always the same. According to "Michael Kay (2008): XSLT 2.0 and XPath 2.0" (4th ed):

_At any point in the processing of a stylesheet, there is generally a context item. The XPath expression '.', if it is used as a standalone expression within a stylesheet, will always select the context item and will return the same result as the expression 'current()'. Within an XPath expression, the context item may change; for example, within a predicate, the context item is the item being tested using the predicate. The result of the current() function, however, is the same wherever it is used within an expression. This makes it useful within predicates, as a way of referring to the item that would have been the context item at the outermost level._

An example: 
<?xml version="1.0" encoding="UTF-8"?>
<test>
  <parts>
    <part code="w1">a</part>
    <part code="w1">b</part>
    <part code="w2">c</part>
    <part code="w2">d</part>
  </parts>
  <wholes>
    <whole code="w1"/>
    <whole code="w2"/>
    <whole code="w3"/>
  </wholes>
</test>

The Schematron file:

<?xml version="1.0" encoding="UTF-8"?>
<sch:schema xmlns:sch="http://purl.oclc.org/dsdl/schematron" queryBinding="xslt2">
  <sch:pattern>
    <sch:rule context="whole">
      <sch:assert test="//part[@code = current()/@code]">A 'whole' must have at least one 'part' with same code.</sch:assert>
    </sch:rule>
  </sch:pattern>
</sch:schema>

Validating the example with the Schematron file results in an error. If you replace 'current()' with '.' then validation will pass without error - but that won't be the intended result.

**Comment by vog** [comment]
This is a bit disappointing, as this forbids the usage of alternative Schematron processors. Not all of them are using the ISO-Schematron implementation based on XSLT(2).

As far as I can see `current()` is only used at the top of the generated statements, such as:

    every $x1 in current() satisfies (...)

Here, the replacement is perfectly valid:

    every $x1 in . satisfies (...)

Moreover, since there is a variable `$x1` bound to that value, this can be used throughout instead of `curent()`. And I believe this is already the case, isn't it?

So to summarize (please correct me if I'm wrong):

* The `current()` statement is only used at top level
* The `current()` statement is not needed at lower levels (`$x1` is used instead)
* Chaging `current()` to `.` is correct at that level, and works for XSLT-based as well as non-XSLT-based Schematron processors (similar to the proposal at #12)

**Comment by jechterhoff** [comment]
Based on what you said, the "context" of this issue is the FOL to Schematron target, not the OCL to Schematron conversion of the XML Schema target. That's an important difference.

In this context, it's true that - at the moment - current() is only used at top level. This is due to what the grammar currently supports.

So, a new conversion rule could be added to the FOL to Schematron target, to use '.' whenever 'current()' would be used.

**Comment by jechterhoff** [comment]
Since there has not been more discussion for over a year, I'm closing this issue.
